### PR TITLE
Scores that don't require a forward prop each

### PR DESCRIPTION
**Pull request by dnouri**
- on_batch_end now also gets 'y_pred' and 'loss'
- Add EpochMetrics and BatchMetrics callbacks

  These make use of the 'y_pred' parameter that NeuralNet now passes
  into callbacks.  Using 'y_pred', it's no longer necessary to do a
  separate forward run of the network just to produce scores, when a
  forward run has already been performed to calculate the loss.

  Refactoring existing code to use this and deprecate BatchScoring and
  EpochScoring is a TODO.  Plus possibly more improvements.

### CODE REVIEW

**Comment by ottonemo** [comment]
We should refrain from adding things to callback parameters that are available through the net's history, like loss for example.

**Comment by dnouri** [comment]
> We should refrain from adding things to callback parameters that are available through the net's history, like loss for example.

Meh.  How about we remove the special case `self.history.record_batch` calls inside the training loop here and move them out into a callback.

**Comment by ottonemo** [comment]
> Meh. How about we remove the special case self.history.record_batch calls inside the training loop here and move them out into a callback.

So every scoring callback adds the value to the history? That would only work if we have some sort of scoring callback bucket that's prefilled with a loss metric and can be extended by the user. Otherwise this would only be a more complicated change.

**Comment by dnouri** [comment]
> So every scoring callback adds the value to the history? That would only work if we have some sort of scoring callback bucket that's prefilled with a loss metric and can be extended by the user. Otherwise this would only be a more complicated change.

Isn't it the case already that the sorting of callbacks is a bit delicate?  I remember someone saying that the `PrintLog` has to come after the other callbacks etc.  I'd suggest we look for a general solution for this kind of dependency problem, e.g. a simple `priority` (int) attribute that callbacks are sorted by before firing.  This way we could ensure that what we want to come first, will come first.  User defined callbacks could default to coming before the `PrintLog` and after whatever else we want to run before.

Also, now that `loss` and `Xi` (and thus `len(Xi)`) are available as parameters in callbacks, it would seem unnecessary for new callbacks to rely on getting those values from the history, yes?

**Comment by ottonemo** [comment]
> Isn't it the case already that the sorting of callbacks is a bit delicate? 

It is and we probably need to address that sooner or later (see #54). I think we don't have to decide on where to move this bit of code here and now. Instead, we should focus on how this PR came to be, what the benefits are and what is still to do.

There are two angles for why we might consider a different scoring approach:
1. support datasets as inputs which break the assumption that we can always provide `X` and `y` with `on_epoch_begin`, this breaks `EpochScoring` but `BatchScoring` is fine
2. re-use of predictions to improve performance when using multiple scores

(1) would benefit from any solution that can switch to a different scoring level (epoch vs. batch) depending on the input data type OR caches all inputs for the whole epoch (probably not wise). This is not directly associated with this PR but should be considered when rewriting the scoring callback. 

(2) is the main reason for this PR and has several implications. First we cannot use sklearn scorers for this as they require the estimator and call `predict`/`predict_proba` depending on the chosen scoring. This makes it hard to introduce caching and any solution (proxy estimator, overwriting `predict`, ...) is probably to messy. So we need to switch to metrics directly (which only depend on `y_true` and `y_pred`). The next problem is that the predictions need to be stored without redundancy (`y_pred` might be quite large) where we have the option to store it in the history object or, as suggested with your change, in a central callback that handles the individual metrics themselves. I see one problem with this approach, namely that it is hard to have default metrics while still being able to add new callbacks, e.g.:
```
class Net:
    default_callbacks = Metrics(['accuracy'])
Net(callbacks=Metrics([....])
```
In this case we would have to override the metrics completely which might be a bit of a pain if there are more complicated metrics supplied by default and the user has to re-type some code from the defaults. While this is probably not a problem it might not be a concern but we should at least discuss it first.

**Comment by dnouri** [comment]
Thanks for your comments.

> There are two angles for why we might consider a different scoring approach:
> 1. support datasets as inputs which break the assumption that we can always provide `X` and `y` with `on_epoch_begin`, this breaks `EpochScoring` but `BatchScoring` is fine
> 2. re-use of predictions to improve performance when using multiple scores

For (2), note that you already get a speed benefit from this patch even when calculating a single score!  Since the first scorer (such as the default `valid_acc`) already incurs an extra forward pass.  And it gets worse as you add scorers.

> (1) would benefit from any solution that can switch to a different scoring level (epoch vs. batch) depending on the input data type OR caches all inputs for the whole epoch (probably not wise). This is not directly associated with this PR but should be considered when rewriting the scoring callback.

I'd prefer having the user control which scoring function to use and refrain from automatic switching, but it's not a strong opinion.

So `BatchMetrics` doesn't need to collect the dataset before it does the scoring, just like `BatchScoring`.  Neither `BatchMetrics` nor `EpochMetrics` depend on the full `X` and `y` at `on_epoch_begin` time, so there's a path to supporting datasets where we don't have the complete `X` and `y` in memory.  (Mentioning this only for clarity.)

If you think keeping all targets and predictions in memory (like `EpochMetrics` does it) is problematic, consider that today we're requiring users of `EpochScoring` to load all training data into memory, so it's not really worse than the status quo; it's keeping all predictions and targets in memory vs. keeping all data and targets in memory.  Of course targets can be as big as inputs, but often the result will be a _smaller_ memory footprint than what we have today.

`BatchMetrics` should be the default for obvious reasons.  User can override with `EpochMetrics` if they care about it, such as when calculating F1.  No danger of blowing people's memory up if they don't change the defaults.

> (2) [...] I see one problem with [your approach of using a single callback for all metrics], namely that it is hard to have default metrics while still being able to add new callbacks, e.g.:
> ```
> class Net:
>     default_callbacks = Metrics(['accuracy'])
> Net(callbacks=Metrics([....])
> ```
> In this case we would have to override the metrics completely which might be a bit of a pain if there are more complicated metrics supplied by default and the user has to re-type some code from the defaults. While this is probably not a problem it might not be a concern but we should at least discuss it first.

Not concerned.  ;-)

Although with this patch it becomes viable to use multiple scores/metrics per default, I don't see us defining lots of default metrics.  (Though I would suggest calculating train _and_ valid accuracy for classification.)

**Comment by ottonemo** [comment]
> I'd prefer having the user control which scoring function to use and refrain from automatic switching, but it's not a strong opinion.

It should definitely be user-controlled but the case epoch metrics + dataset should raise an error.

> So BatchMetrics doesn't need to collect the dataset before it does the scoring, just like BatchScoring. Neither BatchMetrics nor EpochMetrics depend on the full X and y at on_epoch_begin time, so there's a path to supporting datasets where we don't have the complete X and y in memory. (Mentioning this only for clarity.)

Right, this also means that we should probably change the signature of `on_epoch_begin` and `on_epoch_end` to accept a dataset (internal or provided) instead of `X` and `y`. This is more related to #146 but the scoring callback currently depends on `X`/`y` at `on_epoch_end`.

> `BatchMetrics` should be the default for obvious reasons. User can override with `EpochMetrics` if they care about it, such as when calculating F1. No danger of blowing people's memory up if they don't change the defaults.

Not sure about that. If we use batch-level by default then it is very easy to screw up F1 scores. I think we should make the safest setting (in terms of correctness) the default which would be epoch-level scoring.

> Although with this patch it becomes viable to use multiple scores/metrics per default, I don't see us defining lots of default metrics. (Though I would suggest calculating train and valid accuracy for classification by default.)

I'm fine with a solution that requires re-setting the accuracy metrics if the user wants to specify additional metrics.

Another point is that we the `Metrics` callback should support `set_params`, i.e. the user should be able to change aspects of the individual metrics contained in the callback , for example, `callback__metrics__accuracy__lower_is_better=False`.

**Comment by ottonemo** [comment]
> I'm fine with a solution that requires re-setting the accuracy metrics if the user wants to specify additional metrics.

Actually I'm not fine with this unless we find a good way to integrate train and valid loss scoring because those need to be there by default and I would get annoyed rather quickly if I had to specify them all over again when adding new metrics.

**Comment by ottonemo** [comment]
OK I think the default metric problem is not an actual problem. Every network instance will have default `Metrics` for train and valid loss and, in case of classification, could potentially have an accuracy metric. All these metrics are batch-level metrics which don't store any additional information except the metric score. Therefore defining a new `Metrics` instance at network initialization and having two `Metrics` callbacks in the list of active callbacks is most likely not a problem.

**Comment by ottonemo** [comment]
Closing in favor of #171.
